src/MapViewController.swift

```diff
@@ -417,11 +476,184 @@ open class MapViewController: UIViewController, LocationManagerDelegate, TGRecog
+  
+  //MARK : TGMapViewDelegate
+  
+  open func mapView(_ mapView: TGMapViewController, didLoadSceneAsync scene: String) {
+    guard (loadDelegate != nil) && (self.loadDelegate?.responds(to: #selector(MapLoadDelegate.mapController(_:didLoadSceneAsync:))))! else {
+      return
+    }
+    self.loadDelegate?.mapController!(self, didLoadSceneAsync: scene)
+  }
+  
+  open func mapViewDidCompleteLoading(_ mapView: TGMapViewController) {
+    guard (self.loadDelegate != nil) && (self.loadDelegate?.responds(to: #selector(MapLoadDelegate.mapControllerDidCompleteLoading(_:))))! else {
+      return
+    }
+    self.loadDelegate?.mapControllerDidCompleteLoading!(self)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectFeature feature: [AnyHashable : Any]?, atScreenPosition position: CGPoint) {
+    guard (feature != nil) else {
+      return
+    }
+    guard (self.selectDelegate != nil) else {
+      return
+    }
+    guard (self.selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectFeature:atScreenPosition:))))! else {
+      return
+    }
+    self.selectDelegate?.mapController!(self, didSelectFeature: feature, atScreenPosition: position)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectLabel labelPickResult: TGLabelPickResult?, atScreenPosition position: CGPoint) {
+    guard (labelPickResult != nil) else {
+      return
+    }
+    guard (self.selectDelegate != nil) else {
+      return
+    }
+    guard (self.selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectLabel:atScreenPosition:))))! else {
+      return
+    }
+    self.selectDelegate?.mapController!(self, didSelectLabel: labelPickResult, atScreenPosition: position)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectMarker markerPickResult: TGMarkerPickResult?, atScreenPosition position: CGPoint) {
+    guard (markerPickResult != nil) else {
+      return
+    }
+    guard (self.selectDelegate != nil) else {
+      return
+    }
+    guard (self.selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectMarker:atScreenPosition:))))! else {
+      return
+    }
+    self.selectDelegate?.mapController!(self, didSelectMarker: markerPickResult, atScreenPosition: position)
+  }
+  
+  //MARK : TGRecognizerDelegate
+  
+  open func mapView(_ view: TGMapViewController, recognizer: UIGestureRecognizer, shouldRecognizeSingleTapGesture location: CGPoint) -> Bool {
+    guard (self.gestureDelegate != nil) else {
+      return true
+    }
+    guard (self.gestureDelegate?.responds(to: #selector(MapGestureDelegate.mapController(_:recognizer:shouldRecognizeSingleTapGesture:))))! else {
+      return true
+    }
+    let recognize = (self.gestureDelegate?.mapController!(self, recognizer: recognizer, shouldRecognizeSingleTapGesture: location))!
+    if !recognize {
+      self.tgViewController.pickLabel(at: location)
```

Comment:
I realized that the weird behavior I was seeing was mostly due to not checking that the pick result was nil :(